include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,190 @@
+ */
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API CartesianGrid
+    {
+        friend class CartesianGridBuilder< DIMENSION >;
+
+    public:
+        CartesianGrid( intvecn< DIMENSION > nb_cells,
+            index_t nb_total_cells,
+            ReferenceFrame< DIMENSION > vec_cartesian_axis )
+            : nb_cells_( nb_cells ), cartesian_frame_( vec_cartesian_axis )
+        {
+            nb_total_cells_ = 1;
+            for( auto i : RINGMesh::range( DIMENSION ) )
+            {
+                nb_total_cells_ *= nb_cells;
+            }
+        }
+
+        static MeshType type_name_static()
+        {
+            return 'CartesianGrid';
+        }
+        void save_mesh( const std::string& filename ) const
+        {
+            // TODO
+        }
+        void print_mesh_bounded_attributes() const
+        {
+            // TODO
+        }
+        //		GEO::AttributesManager& vertex_attribute_manager() const
+        // override
+        //		{
+        //			return mesh_->vertices.attributes();
+        //		}
+        MeshType type_name() const override
+        {
+            return type_name_static();
+        }
+        //		static std::string default_extension_static()
+        //		{
+        //			return "geogram";
+        //		}
+        //		std::string default_extension() const override
+        //		{
+        //			return default_extension_static();
+        //		}
+        //		index_t nb_vertices() const override
+        //		{
+        //			return mesh_->vertices.nb();
+        //		}
+
+    private:
+        /**
+         * \brief Forbids copy.
+         * \details This is to make sure that client code does
+         *   not unintentionally copy a Mesh (for
+         *   instance by passing it by-value to a function).
+         *   Use copy() instead.
+         */
+        CartesianGrid( const CartesianGrid& rhs );
+
+        /**
+         * \brief Forbids copy.
+         * \details This is to make sure that client code does
+         *   not unintentionally copies a Mesh (for
+         *   instance by passing it by-value to a function).
+         *   Use copy() instead.
+         */
+        const CartesianGrid& operator=( const CartesianGrid& rhs );
+
+    protected:
+        intvecn< DIMENSION > nb_cells_;
+        index_t nb_total_cells_;
+
+        ReferenceFrame< DIMENSION > cartesian_frame_;
+
+    public:
+        vecn< DIMENSION > cell_center_reference_coords(
+            const intvecn< DIMENSION > cartesian_coords ) const
+        {
+            vecn< DIMENSION > cart_double_coords;
+            for( auto i : RINGMesh::range( DIMENSION ) )
+            {
+                cart_double_coords[i] =
+                    static_cast< double >( cartesian_coords[i] );
+            }
+            return cartesian_frame_.coords_to_base( cart_double_coords );
+        }
+
+        intvecn< DIMENSION > containing_cell(
+            const vecn< DIMENSION > vertex ) const
```

Comment:
You can do this operation on point not only on vertex. And in the name you can precise that you are working with local coordinates